omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,790 @@
+    q.enqueue_chunk(&start, 1);
+    levels[start] = 0;
+
+    // Used for synchronization of (pseudo)remaining work.
+    int threads_working = 0;
+
+#pragma omp parallel firstprivate(max_degree)
+    {
+        // This is an upper bound, as '.deque_chunk...' is upper bounded by
+        // chunk_bound.
+        vector<IndexType> local_to_insert(max_degree * chunk_bound, exec);
+        size_type local_to_insert_size = 0;
+
+        while (true) {
+            // Get some nodes from the locked q.
+            const auto pair = q.dequeue_chunk(&threads_working);
+            const auto chunk = pair.first;
+            const auto chunk_size = pair.second;
+
+            // If there are nodes left to process.
+            // If there are no nodes returned that also means there are no nodes
+            // still in flight.
+            if (chunk_size > 0) {
+                for (IndexType chunk_i = 0; chunk_i < chunk_size; ++chunk_i) {
+                    const auto node = chunk[chunk_i];
+
+                    // For each node in the chunk, process all neighbours,
+                    // by reducing their level if a new shortest path was found,
+                    // in that case also writing the neighbour into
+                    // local_to_insert.
+                    reduce_neighbours_levels(
+                        num_vertices, row_ptrs, col_idxs, &local_to_insert[0],
+                        &local_to_insert_size, levels, node);
+                }
+
+                // Insert back to global work queue, done with this chunk.
+                q.enqueue_chunk(&local_to_insert[0], local_to_insert_size);
+
+#pragma omp atomic update
+                --threads_working;
+
+                local_to_insert_size = 0;
+            } else {
+                // Done!
+                break;
+            }
+        }
+    }
+}
+
+
+/**
+ * Finds a 'contender', meaning a node in the last level of the rls with minimum
+ * degree, returns it along with the rls height.
+ */
+template <typename IndexType>
+std::pair<IndexType, IndexType> rls_contender_and_height(
+    std::shared_ptr<const OmpExecutor> exec, const IndexType num_vertices,
+    const IndexType *const row_ptrs, const IndexType *const col_idxs,
+    const IndexType *const degrees,
+    IndexType *const levels,  // Must be max/inf in all nodes connected to start
+    const IndexType start, const IndexType max_degree)
+{
+    // Create a level structure.
+    ubfs(exec, num_vertices, row_ptrs, col_idxs, levels, start, max_degree);
+
+    // Find a node in the last level with minimal degree, the "contender".
+
+#pragma omp declare reduction(find_contender \
+                              : std::pair <std::pair<IndexType,IndexType>, IndexType> \
+                              : omp_out = omp_in.first.first > omp_out.first.first ? omp_in : omp_in.first.first == omp_out.first.first \
+                              && omp_in.first.second > omp_out.first.second ? omp_in : omp_out)
```

Comment:
Can you try and hand-format this to a maximum width of 80 characters?
Looks like `clang-format` does not want to touch this.

Reply:
Can you not use the operator< on std::pair to simplify this a bit?

Reply:
@upsj That actually works. Learning something new everyday :)